fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -1351,3 +1364,193 @@ void TracingOptionsImpl::clear(ReadYourWritesTransaction* ryw, const KeyRef& key
 	ryw->setSpecialKeySpaceErrorMsg("clear disabled");
 	throw special_keys_api_failure();
 }
+
+CoordinatorsImpl::CoordinatorsImpl(KeyRangeRef kr) : SpecialKeyRangeRWImpl(kr) {}
+
+Future<Standalone<RangeResultRef>> CoordinatorsImpl::getRange(ReadYourWritesTransaction* ryw, KeyRangeRef kr) const {
+	Standalone<RangeResultRef> result;
+	KeyRef prefix(getKeyRange().begin);
+	// the constructor of ClusterConnectionFile already checks whether the file is valid
+	auto cs = ClusterConnectionFile(ryw->getDatabase()->getConnectionFile()->getFilename()).getConnectionString();
+	auto coordinator_processes = cs.coordinators();
+	Key cluster_decription_key = prefix.withSuffix(LiteralStringRef("cluster_description"));
+	if (kr.contains(cluster_decription_key)) {
+		result.push_back_deep(result.arena(), KeyValueRef(cluster_decription_key, cs.clusterKeyName()));
+	}
+	// Note : the sort by string is anti intuition, ex. 1.1.1.1:11 < 1.1.1.1:5
+	// include :tls in keys if the network addresss is TLS
+	std::sort(coordinator_processes.begin(), coordinator_processes.end(),
+	          [](const NetworkAddress& lhs, const NetworkAddress& rhs) { return lhs.toString() < rhs.toString(); });
+	std::string processes_str;
+	for (auto& w : coordinator_processes) {
+		if (processes_str.size()) processes_str += ",";
+		processes_str += w.toString();
+	}
+	Key processes_key = prefix.withSuffix(LiteralStringRef("processes"));
+	if (kr.contains(processes_key)) {
+		result.push_back_deep(result.arena(), KeyValueRef(processes_key, Value(processes_str)));
+	}
+	return rywGetRange(ryw, kr, result);
+}
+
+ACTOR static Future<Optional<std::string>> coordinatorsCommitActor(ReadYourWritesTransaction* ryw, KeyRangeRef kr) {
+	state Reference<IQuorumChange> change;
+	state std::vector<NetworkAddress> addressesVec;
+	state std::vector<std::string> process_address_strs;
+	state Optional<std::string> msg;
+	state int index;
+	state bool parse_error = false;
+
+	// check update for cluster_description
+	Key processes_key = LiteralStringRef("processes").withPrefix(kr.begin);
+	auto processes_entry = ryw->getSpecialKeySpaceWriteMap()[processes_key];
```

Comment:
Not sure if I understand correctly, I think not, since it is `KeyRangeMap`, where [[]](https://github.com/apple/foundationdb/blob/73e77f31eb31ae8bbac826e8534790613eb8d751/fdbrpc/RangeMap.h#L122) only returns the corresponding value of the range the key belongs to.